src/shell/stage-in.c

```diff
@@ -0,0 +1,477 @@
+            json_array_foreach (blobvec, index, o) {
+                if (extract_blob (ctx->h, archive, path, o) < 0)
+                    goto error;
+            }
+        }
+    }
+    archive_entry_free (entry);
+    return 0;
+error:
+    archive_entry_free (entry);
+    return -1;
+
+}
+
+static int extract (struct stage_in *ctx, struct archive *archive)
+{
+    flux_future_t *f;
+    size_t index;
+    json_t *entry;
+
+    if (!(f = mmap_list (ctx->h,
+                         ctx->direct ? false : true,
+                         ctx->tags,
+                         ctx->pattern))) {
+        shell_log_error ("mmap-list: %s", strerror (errno));
+        return -1;
+    }
+    for (;;) {
+        json_t *files;
+
+        if (flux_rpc_get_unpack (f, "{s:o}", "files", &files) < 0) {
+            if (errno == ENODATA)
+                break; // end of stream
+            shell_log_error ("mmap-list: %s", future_strerror (f, errno));
+            return -1;
+        }
+        json_array_foreach (files, index, entry) {
+            if (ctx->direct) {
+                if (extract_file (ctx, archive, entry) < 0)
+                    return -1;
+            }
+            else {
+                json_t *fileref;
+
+                if (!(fileref = load_fileref (ctx->h,
+                                              json_string_value (entry))))
+                    return -1;
+                if (extract_file (ctx, archive, fileref) < 0) {
+                    json_decref (fileref);
+                    return -1;
+                }
+                json_decref (fileref);
+            }
+            ctx->count++;
+        }
+        flux_future_reset (f);
+    }
+    flux_future_destroy (f);
+    return 0;
+}
+
+static int extract_files (struct stage_in *ctx)
+{
+    char *orig_dir;
+    struct archive *archive = NULL;
+    struct timespec t;
+
+    if (!(orig_dir = getcwd (NULL, 0))) {
+        shell_log_error ("getcwd: %s", strerror (errno));
+        return -1;
+    }
+    if (!(archive = archive_write_disk_new ())) {
+        shell_log_error ("error creating libarchive context");
+        goto error;
+    }
+    if (chdir (ctx->destdir) < 0) {
+        shell_log_error ("chdir %s: %s", ctx->destdir, strerror (errno));
+        goto error;
+    }
+    shell_debug ("=> %s", ctx->destdir);
+    monotime (&t);
+    if (extract (ctx, archive) < 0)
+        goto error ;
+    double elapsed = monotime_since (t) / 1000;
+    shell_debug ("%d files %.1fMB/s",
+                 ctx->count,
+                 1E-6 * ctx->total_size / elapsed);
+    if (chdir (orig_dir) < 0) {
+        shell_die (1,
+                   "could not chdir back to original directory %s: %s",
+                   orig_dir,
+                   strerror (errno));
+        goto error;
+    }
+    archive_write_free (archive);
+    free (orig_dir);
+    return 0;
+error:
+    if (archive)
+        archive_write_free (archive);
+    free (orig_dir);
+    return -1;
+}
```

Comment:
should try to `chdir()` to orig dir on error, in case caller wants to survive the error?